### ISSUE TITLE
Native federation can not be added in ESM module

### ISSUE THREAD

**Issue by ZissisT** [angular-architects/module-federation-plugin#681]
### With what library do you have an issue?

native-federation

### Reproduction of the bug/regression with instructions

In a new Angular18+ project that has "type": "module" in its package.json, native federation can not be initialized because it uses require() and hardcoded 'federation.config.js' file name for initialization. Exception is thrown, asking to rename .js file to .cjs.

I changed hardcoded file name in libs/native-federation/src/builders/build/builder.ts  from 'federation.config.js' to 'federation.config.cjs' and worked like a charm. I can push a PR to check for .cjs first and then for .js, if needed (or even have an optional filename parameter for config file)

Here is a Stackblitz example that throws this exception:
[https://stackblitz.com/edit/stackblitz-starters-1ggp6r](https://stackblitz.com/edit/stackblitz-starters-1ggp6r)

### Expected behavior

Expected to load with no error. If an Angular lib/app can not be initialized as ESM module due to native federation then what?
We can not change our app's nature, to commonjs just for this, it leads to a chain of changes that make this unusable.

### Versions of Native/Module Federation, Angular, Node, Browser, and operating system

"@angular-architects/native-federation": "18.2.2"
Angular CLI: 18.2.10
Node: 20.13.1
OS: linux x64
Browser: Chrome 130.0.6723.70 (Official Build) (64-bit)

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by manfredsteyer** [comment]
Good Point. Thanks for this. It seems like the issue only comes up if people add `type: "module"` to their `package.json`. I think we need several things here:

- Respect the new file name with .cjs when loading the federation config
- Generating the right filename in the ng add schematic
- Updating the file name in an ng update schematic
- displaying an error message on the cnsole when we see an old federation.config.js

PRs are welcomed. Please link the Pr here.

**Comment by ZissisT** [comment]
@manfredsteyer PR is ready and I am trying to test it, but schematics seem broken, I get "path" undefined errors constantly after using ng add to add host/remotes.
Reason is this:
module-federation-plugin/libs/native-federation/src/schematics/init/schematic.ts , line 352
`  const main =
    projectConfig.architect.build.options.main ||
    projectConfig.architect.build.options.browser`
    
    which are not supported any more in the native-federation builders schematic but exist in the esbuild section of angular.json
    Changing to/adding
     `projectConfig.architect.esbuild.options.main || projectConfig.architect.esbuild.options.main` 
     
    fixes everything. Should I raise a new issue for this? Or add fix in this one?
    
    Thank you

**Comment by manfredsteyer** [comment]
Awesome. Big thanks. I plan for another release at the End of Nov. I will look into this PR when making this release ready.

**Comment by ZissisT** [comment]
@manfredsteyer what about the update requirement? I haven't created a new update schematic but added a new migration. Is this correct or should I add a new update schematic that updates the name? Ty

**Comment by ms-dosx86** [comment]
Any updates?

**Comment by Ionaru** [comment]
Awaiting the merge & release of ESM support, I've made a small `postinstall` script to patch `@angular-architects/native-federation` and renamed `federation.config.js` to `federation.config.cjs`

```js
import { readFileSync, writeFileSync } from "node:fs";

/**
 * Patches @angular-architects/native-federation to find our commonjs federation config.
 */

console.log("Starting patch of @angular-architects/native-federation...");

const filePath = "node_modules/@angular-architects/native-federation/src/builders/build/builder.js";
const content = readFileSync(filePath, "utf-8");
const updatedContent = content.replaceAll("federation.config.js", "federation.config.cjs");
writeFileSync(filePath, updatedContent);

console.log("@angular-architects/native-federation patched.");
```

---

Support for writing `federation.config.js` in ESM syntax would also be welcome, but that fails within `@softarc/native-federation/build` due to the resolved config having a nested `default` object.

It fails here: https://github.com/angular-architects/module-federation-plugin/blob/33393070c0ec83e188589c1a5b13bb0b63a42d66/libs/native-federation/src/builders/build/builder.ts#L178-L192

```
config [Module: null prototype] {
  __esModule: true,
  default: {
    name: 'name-of-project',
    exposes: {
      './name-of-project': './packages/name-of-project/src/bootstrap.ts'
    },
    shared: {},
    sharedMappings: [
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object]
    ],
    skip: { strings: [Set], functions: [], regexps: [Array] },
    externals: [],
    features: { mappingVersion: false }
  }
}
```

**Comment by ZissisT** [comment]
@manfredsteyer or anyone with proper rights need to review this, but it is a bit outdated now I guess, the PR
